QoL improvements for torch.testing.make_tensor

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96334
* #96333
* #96124
* #97010
* #96332
* #97017
* __->__ #96125
* #96331

Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

pmeier committed:
QoL improvements for torch.testing.make_tensor

[ghstack-poisoned]

pmeier commented on torch/testing/_creation.py:
```diff
@@ -14,12 +14,9 @@
     torch.complex64: torch.float32,
     torch.complex128: torch.float64,
 }
-float_to_corresponding_complex_type_map = {
```
This was never used anywhere.

pmeier replied:
Correction: this is used in two other modules. However, since we no longer have the original map, I put this into `common_dtype`.

pmeier commented on torch/testing/_creation.py:
```diff
@@ -14,12 +14,9 @@
-}
 
 
-def _uniform_random(t: torch.Tensor, low: float, high: float):
```
In accordance with other inplace ops, I postfixed this with an underscore.

mruberry replied:
Sure

pmeier commented on torch/testing/_creation.py:
```diff
@@ -73,20 +70,21 @@ def make_tensor(
-        ValueError: if ``requires_grad=True`` is passed for integral `dtype`
-        ValueError: If ``low > high``.
+        ValueError: If ``requires_grad=True`` is passed for integral `dtype`
+        ValueError: If ``low >= high``.
```
We allow `low == high`:

https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L120-L121

mruberry replied:
OK but shouldn't this stay `low > high` then? Because this suggests that a `ValueError` will be raised when `low == high`

pmeier replied:
:facepalm: Yes, you are right. Let me revert that.

pmeier commented on torch/testing/_creation.py:
```diff
@@ -103,27 +101,47 @@ def make_tensor(
     """
 
-    def _modify_low_high(low, high, lowest, highest, default_low, default_high, dtype):
+    def modify_low_high(
```
No need to make this "private", i.e. leading underscore, since this is a local function and cannot be called from anywhere else.

pmeier commented on torch/testing/_creation.py:
```diff
@@ -103,27 +101,47 @@ def make_tensor(
+    def modify_low_high(
+        low: Optional[float],
+        high: Optional[float],
+        *,
```
Having `low` and `high` as positional is fine for a function called `modify_low_high`. Everything else should be keyword only to enforce readable calls. See below for the result.

pmeier commented on torch/testing/_creation.py:
```diff
@@ -135,53 +153,72 @@ def clamp(a, l, h):
         shape = shape[0]  # type: ignore[assignment]
     shape = cast(Tuple[int, ...], tuple(shape))
 
+    if noncontiguous and memory_format is not None:
+        raise ValueError(
```
This was a plain `assert` before.

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

mruberry commented on torch/testing/_creation.py:
```diff
@@ -135,53 +153,72 @@ def clamp(a, l, h):
+    if noncontiguous and memory_format is not None:
+        raise ValueError(
+            f"The parameters `noncontiguous` and `memory_format` are mutually exclusive, "
+            f"but got {noncontiguous=} and {memory_format=}"
```
I didn't know writing `noncontiguous=` like that would work.

Is it equivalent to `f"noncontiguous={noncontiguous}"`?

pmeier replied:
Yes, that is equivalent. This is an addition to Python 3.8, which is the lowest supported version for PyTorch as of today. See https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging

mruberry commented on torch/testing/_creation.py:
```diff
@@ -135,53 +153,72 @@ def clamp(a, l, h):
+                high,
+                lowest=torch.iinfo(dtype).min,
+                highest=torch.iinfo(dtype).max,
+                # This is incorrect for `torch.uint8`, but since we clamp to `lowest`, i.e. 0 for `torch.uint8`,
```
Nice comment

mruberry reviewed (approved):
<img src="https://media1.giphy.com/media/1034EEGrn91SrS/giphy.gif"/>

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier commented on torch/testing/_creation.py:
```diff
@@ -8,18 +8,14 @@
 import torch
 
-# Used by make_tensor for generating complex tensor.
-complex_to_corresponding_float_type_map = {
```
We don't need that since `torch.finfo` works for complex types as well.

pmeier commented on torch/testing/_creation.py:
```diff
@@ -8,18 +8,14 @@
-
-
-def _uniform_random(t: torch.Tensor, low: float, high: float):
+_INTEGRAL_TYPES = [torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64]
+_FLOATING_TYPES = [torch.float16, torch.bfloat16, torch.float32, torch.float64]
+_COMPLEX_TYPES = [torch.complex32, torch.complex64, torch.complex128]
+_BOOLEAN_OR_INTEGRAL_TYPES = [torch.bool, *_INTEGRAL_TYPES]
+_FLOATING_OR_COMPLEX_TYPES = [*_FLOATING_TYPES, *_COMPLEX_TYPES]
```
Defining them here do avoid doing it over and over again inside `make_tensor`.

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

mruberry reviewed (approved):
<img src="https://media1.giphy.com/media/10RgsuetO4uDkY/giphy.gif"/>

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]

pmeier committed:
Update on "QoL improvements for torch.testing.make_tensor"


Per title. The major ones:

- Enforce keyword only parameters for `_modify_low_high`, which takes 7 parameters.
  https://github.com/pytorch/pytorch/blob/28aa2efd1422e689d3a7dc6b447ca1f31bbad1c3/torch/testing/_creation.py#L147
  is just impossible to comprehend without multiple trips back and forth.
- Improve the error messages by including the offending values in the message

I'll highlight the smaller ones inline.

[ghstack-poisoned]
